worldedit-bukkit/adapters/adapter-1.17.1/src/main/java/com/sk89q/worldedit/bukkit/adapter/impl/v1_17_R1_2/PaperweightAdapter.java

```diff
@@ -500,28 +500,34 @@ public Component getRichItemName(BaseItemStack itemStack) {
     }
 
     @SuppressWarnings({ "unchecked", "rawtypes" })
-    @Override
-    public Map<String, ? extends Property<?>> getProperties(BlockType blockType) {
-        Map<String, Property<?>> properties = Maps.newTreeMap(String::compareTo);
-        Block block = getBlockFromType(blockType);
-        StateDefinition<Block, net.minecraft.world.level.block.state.BlockState> blockStateList =
-            block.getStateDefinition();
-        for (net.minecraft.world.level.block.state.properties.Property state : blockStateList.getProperties()) {
-            Property property;
+    private static final LoadingCache<net.minecraft.world.level.block.state.properties.Property, Property<?>> propertyCache = CacheBuilder.newBuilder().build(new CacheLoader<net.minecraft.world.level.block.state.properties.Property, Property<?>>() {
+        @Override
+        public Property<?> load(net.minecraft.world.level.block.state.properties.Property state) throws Exception {
             if (state instanceof net.minecraft.world.level.block.state.properties.BooleanProperty) {
-                property = new BooleanProperty(state.getName(), ImmutableList.copyOf(state.getPossibleValues()));
+                return new BooleanProperty(state.getName(), ImmutableList.copyOf(state.getPossibleValues()));
             } else if (state instanceof DirectionProperty) {
-                property = new DirectionalProperty(state.getName(),
+                return new DirectionalProperty(state.getName(),
                     (List<Direction>) state.getPossibleValues().stream().map(e -> Direction.valueOf(((StringRepresentable) e).getSerializedName().toUpperCase(Locale.ROOT))).collect(Collectors.toList()));
             } else if (state instanceof net.minecraft.world.level.block.state.properties.EnumProperty) {
-                property = new EnumProperty(state.getName(),
+                return new EnumProperty(state.getName(),
                     (List<String>) state.getPossibleValues().stream().map(e -> ((StringRepresentable) e).getSerializedName()).collect(Collectors.toList()));
             } else if (state instanceof net.minecraft.world.level.block.state.properties.IntegerProperty) {
-                property = new IntegerProperty(state.getName(), ImmutableList.copyOf(state.getPossibleValues()));
+                return new IntegerProperty(state.getName(), ImmutableList.copyOf(state.getPossibleValues()));
             } else {
                 throw new IllegalArgumentException("WorldEdit needs an update to support " + state.getClass().getSimpleName());
             }
+        }
+    });
 
+    @SuppressWarnings({ "rawtypes" })
+    @Override
+    public Map<String, ? extends Property<?>> getProperties(BlockType blockType) {
+        Map<String, Property<?>> properties = Maps.newTreeMap(String::compareTo);
```

Comment:
Strings are natively comparable. We don't need the extra indirection. Also, we don't need `Maps`.
```suggestion
        Map<String, Property<?>> properties = new TreeMap<>();
```